### ISSUE TITLE
[feature] don't install empty directories

### ISSUE THREAD

**Issue by sadsumi** [deadc0de6/dotdrop#314]
My use case might be too specific to me but, I don't know, someone might find it useful. It's also more of a cosmetic change.

**What I need**

If there's an empty directory managed by dotdrop (either because it's actually empty or it/its content is ignored), I don't want it to be copied when installing, possibly by setting some `ignoreemptydir` option. `dotdrop install` should also never mention it in its output.

**What I've tried**

First of all, I think it's currently not true. If I have an `awesome` directory that is managed by dotdrop, and then some `awesome/empty` directory in it, even though I add `awesome/empty` to `instignore` (as `*empty` as suggested in docs, or as `*empty*`) it's still gonna be copied. Example:

```yaml
  d_awesome:
    dst: ~/.config/awesome
    src: config/awesome
    link: nolink
    ignoreempty: true
    instignore:
    - '*empty'
```

**Use case**

some of my dotfiles are complex, e.g. `awesome`. It consists of a lot of scripts which use templating to manage different profiles. I add those dotfiles to separate git repositories to manage multiple versions of them. Now, I want to add `.git` directory to `instignore` (`*.git*` to also ignore its content) so it will never be copied to the actual machine and won't clutter the output of install command - which is a bit of a hassle. But despite this setting, empty directories are still created (and `git` creates a lot of directories). They also change whenever I commit or do other things.

I might use an action to delete this directory after it's installed but it still clutters the output and ~~forces me to confirm every single empty directory (I don't force my installs)~~

**Comment by deadc0de6** [comment]
The dotfile entry `ignoreempty` applies to templates (as mentioned [in the doc](https://dotdrop.readthedocs.io/en/latest/config-format/#dotfiles-entry)). The reason dotdrop is able to do that is because it will first transform the content (using templating) and then be able to establish that the content is empty and thus ignore the file creation.

Directories (and more generally files hierarchy) is a different story since these are created before processing their content (files within). I supposed having such a config as `ignoreemptydir` would apply the same reasoning you mentioned above: after installing something, check if any directory/sub-directory is empty and delete it. I guess this would be much easier than processing each and every directory creation.

What about using a simple action like the following (applied to the dotfile for example, *untested*):
```yaml
actions:
  clearemptydir: find '{0}' -empty -type d -delete
dotfiles:
  d_awesome:
      dst: ~/.config/awesome
      src: config/awesome
      link: nolink
      ignoreempty: true
      instignore:
      - '*empty'
      actions:
      - clearemptydir "~/.config/awesome"
```

**Comment by sadsumi** [comment]
Thanks for response, this is certainly a job for an action but I have some issues with this solution which is why I decided to create this issue.

First, I found out that actions are executed only once per dotfile installation (correct me if I'm wrong) which makes sense but won't help in this scenario. Any subsequent `dotdrop install` to update machine's dotfiles will result in creating empty directories again.

Also, this doesn't help with output being cluttered with `-> create directory <empty_dir>`.

inb4 sorry if I'm being picky, this is not a breaking issue but something that I wanted to sort out somehow to increase readability of my setup.

**Comment by deadc0de6** [comment]
The documentation might be misleading with `Note that dotfile actions are only executed when the dotfile is installed.`. What I meant is that actions are executed each time a dotfile is deployed (when the version in dotdrop differs from the one on the filesystem). This means that the above action would be executed on each time `d_awesome` is to be processed (change in contents, new files/dirs, etc).

Regarding the `create directory <dir>`, what you want is more control over the logs? I could potentially remove any sub info (like `chmod ...`, `creating dir ...`, `linked ...`, `install ... to ...`, etc) with a config option.

**Comment by sadsumi** [comment]
> What I meant is that actions are executed each time a dotfile is deployed (when the version in dotdrop differs from the one on the filesystem).

You are right. However, in case of empty directories those actions are not always triggered. Example: If I install for the first time, this action is triggered and empty directories are removed. After second install (without any prior changes to files), empty directories are installed again but this time action is not triggered, diff likely ignores them as it probably focuses on files.

> Regarding the `create directory <dir>`, what you want is more control over the logs?

Ideally, I don't want the creation of empty directories to be reported **if I actually ignored them**. However, from what I can see this might be tricky to implement as when you create the directory you can't tell if they'll end up empty. Being able to disable `creating dir...` logs would be a simple and useful compromise.

Also, I've been worndering about my issue and I'm not entirely sure how dotdrop works internally (might want to get into it one day) but it seems to me that ignore patterns are only applied to files and not directories, is that correct? That would make sense as if I want to ignore a directory, I have to use `*.git*` and not `*.git` or `*.git/`. Would it be possible to apply those ignore patterns to directories as well? If a directory would match ignore pattern it wouldn't be created and processed in the first place. Just a random idea that came to my mind.

**Comment by deadc0de6** [comment]
I see now the problem. I'll have a look and come up with something more robust then. It might take a bit of time though.
I'll also look at the ignore patterns against directories. I agree this should also work.
I'll keep you posted.

**Comment by deadc0de6** [comment]
@sadsumi I have something regarding the creation of empty dir (the *not-creation* actually ;-) ). I tried to create some tests to cover the different use cases but it's pretty tough to be exhaustive. I would therefore be very grateful if you could test it out.

The new version lives in the [emptydir branch](https://github.com/deadc0de6/dotdrop/tree/emptydir). 

I'll see about ignoring directories in the next step.

**Comment by sadsumi** [comment]
Thanks! I've been pretty busy lately but I'll check it out as soon as I can and let you know.

**Comment by sadsumi** [comment]
So I gave it a go directly with my dotfiles and dotdrop config. It works as expected (no empty directories created and no additional output) in both cases I can think of: empty directories (ignored or not) and ignored non-empty directories (like my `.git`). Also, I tried to add an exception like `"!*.git/config"` but it works in this case too (only this specific file installed).

If you mean applying ignore patterns to directories, I think it's not strictly necessary at this point - `*.git/*` just does it for me. It could be good for one thing though - making it harder to confuse what's actually ignored. I know it can be deduced from docs that ignore patterns apply to files but I somehow kept forgetting about this. :wink:
